extract() function doesn't respect active time zone

When no timezone is active:
```sql
> CREATE TABLE "timezones_event" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "dt" timestamptz NOT NULL);
> INSERT INTO "timezones_event" ("dt") VALUES ('2011-01-01T01:30:00+00:00'::timestamptz);
> SELECT dt, EXTRACT(hour FROM dt) FROM timezones_event;
             dt             | extract  
+---------------------------+---------+
  2011-01-01 01:30:00+00:00 |       1  
```
But when a timezone is active ('Africa/Nairobi' is UTC+3), the extracted value isn't localized:
```sql
> SET TIME ZONE 'Africa/Nairobi';
> SELECT dt, EXTRACT(hour FROM dt) FROM timezones_event;
             dt             | extract  
+---------------------------+---------+
  2011-01-01 04:30:00+03:00 |       1  
(1 row)
```
I believe this is a bug in cockroachdb.

The relevant Django test:

```
======================================================================
FAIL: test_extract_func (db_functions.datetime.test_extract_trunc.DateFunctionTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/db_functions/datetime/test_extract_trunc.py", line 204, in test_extract_func
    lambda m: (m.start_datetime, m.extracted)
  File "/home/tim/code/django/django/test/testcases.py", line 1056, in assertQuerysetEqual
    return self.assertEqual(list(items), values, msg=msg)
AssertionError: Lists differ: [(dat[42 chars]1), 19), (datetime.datetime(2016, 6, 15, 14, 10, 50, 123), 19)] != [(dat[42 chars]1), 14), (datetime.datetime(2016, 6, 15, 14, 10, 50, 123), 14)]

First differing element 0:
(datetime.datetime(2015, 6, 15, 14, 30, 50, 321), 19)
(datetime.datetime(2015, 6, 15, 14, 30, 50, 321), 14)

- [(datetime.datetime(2015, 6, 15, 14, 30, 50, 321), 19),
?                                                     ^

+ [(datetime.datetime(2015, 6, 15, 14, 30, 50, 321), 14),
?                                                     ^

-  (datetime.datetime(2016, 6, 15, 14, 10, 50, 123), 19)]
?                                                     ^

+  (datetime.datetime(2016, 6, 15, 14, 10, 50, 123), 14)]
?                                                     ^
```

--- a/docs/generated/sql/functions.md
+++ b/docs/generated/sql/functions.md
@@ -364,7 +364,8 @@ hour, minute, second, millisecond, microsecond, epoch</p>
 <tr><td><a name="extract"></a><code>extract(element: <a href="string.html">string</a>, input: <a href="timestamp.html">timestamptz</a>) &rarr; <a href="float.html">float</a></code></td><td><span class="funcdesc"><p>Extracts <code>element</code> from <code>input</code>.</p>
 <p>Compatible elements: millennium, century, decade, year, isoyear,
 quarter, month, week, dayofweek, isodow, dayofyear, julian,
-hour, minute, second, millisecond, microsecond, epoch</p>
+hour, minute, second, millisecond, microsecond, epoch,
+timezone, timezone_hour, timezone_minute</p>
 </span></td></tr>
 <tr><td><a name="extract"></a><code>extract(element: <a href="string.html">string</a>, input: timetz) &rarr; <a href="float.html">float</a></code></td><td><span class="funcdesc"><p>Extracts <code>element</code> from <code>input</code>.</p>
 <p>Compatible elements: hour, minute, second, millisecond, microsecond, epoch,

--- a/pkg/sql/logictest/testdata/logic_test/datetime
+++ b/pkg/sql/logictest/testdata/logic_test/datetime
@@ -623,12 +623,12 @@ SELECT extract(nansecond from '2001-04-10 12:04:59.34565423'::timestamptz)
 query R
 SELECT extract(hour from '2016-02-10 19:46:33.306157519-04'::timestamptz)
 ----
-19
+23
 
 query R
 SELECT extract(hours from '2016-02-10 19:46:33.306157519-04'::timestamptz)
 ----
-19
+23
 
 query ITTBT
 SELECT k, element, input, date_trunc(element, input::timestamp) = date_trunc_result, date_trunc(element, input::timestamp)::string

--- a/pkg/sql/logictest/testdata/logic_test/timestamp
+++ b/pkg/sql/logictest/testdata/logic_test/timestamp
@@ -149,3 +149,41 @@ select column_name, data_type FROM [SHOW COLUMNS FROM timestamp_test] ORDER BY c
 id   INT8
 t    TIMESTAMP
 ttz  TIMESTAMPTZ(5)
+
+subtest regression_django-cockroachdb_47
+
+statement ok
+SET TIME ZONE -3
+
+query R
+SELECT extract(hour FROM '2001-01-01 13:00:00+01'::timestamptz)
+----
+9
+
+query R
+SELECT extract(hour FROM '2001-01-01 13:00:00'::timestamp)
+----
+13
+
+query R
+SELECT extract(timezone FROM '2001-01-01 13:00:00+01:15'::timestamptz)
+----
+-10800
+
+statement ok
+SET TIME ZONE +3
+
+query R
+SELECT extract(hour FROM '2001-01-01 13:00:00+01'::timestamptz)
+----
+15
+
+query R
+SELECT extract(hour FROM '2001-01-01 13:00:00'::timestamp)
+----
+13
+
+query R
+SELECT extract(timezone FROM '2001-01-01 13:00:00+01:15'::timestamptz)
+----
+10800

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -1805,12 +1805,13 @@ may increase either contention or retry errors, or both.`,
 			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
 				fromTSTZ := args[1].(*tree.DTimestampTZ)
 				timeSpan := strings.ToLower(string(tree.MustBeDString(args[0])))
-				return extractStringFromTimestamp(ctx, fromTSTZ.Time, timeSpan)
+				return extractStringFromTimestampTZ(ctx, fromTSTZ.Time.In(ctx.GetLocation()), timeSpan)
 			},
 			Info: "Extracts `element` from `input`.\n\n" +
 				"Compatible elements: millennium, century, decade, year, isoyear,\n" +
 				"quarter, month, week, dayofweek, isodow, dayofyear, julian,\n" +
-				"hour, minute, second, millisecond, microsecond, epoch",
+				"hour, minute, second, millisecond, microsecond, epoch,\n" +
+				"timezone, timezone_hour, timezone_minute",
 		},
 		tree.Overload{
 			Types:      tree.ArgTypes{{"element", types.String}, {"input", types.Time}},
@@ -4502,16 +4503,25 @@ func extractStringFromTime(fromTime *tree.DTime, timeSpan string) (tree.Datum, e
 	return extractStringFromTimeOfDay(t, timeSpan)
 }
 
-func extractStringFromTimeTZ(fromTime *tree.DTimeTZ, timeSpan string) (tree.Datum, error) {
+func extractTimezoneFromOffset(offsetSecs int32, timeSpan string) tree.Datum {
 	switch timeSpan {
 	case "timezone":
-		return tree.NewDFloat(tree.DFloat(float64(-fromTime.OffsetSecs))), nil
+		return tree.NewDFloat(tree.DFloat(float64(offsetSecs)))
 	case "timezone_hour", "timezone_hours":
-		numHours := -fromTime.OffsetSecs / secsPerHour
-		return tree.NewDFloat(tree.DFloat(float64(numHours))), nil
+		numHours := offsetSecs / secsPerHour
+		return tree.NewDFloat(tree.DFloat(float64(numHours)))
 	case "timezone_minute", "timezone_minutes":
-		numMinutes := -fromTime.OffsetSecs / secsPerMinute
-		return tree.NewDFloat(tree.DFloat(float64(numMinutes % 60))), nil
+		numMinutes := offsetSecs / secsPerMinute
+		return tree.NewDFloat(tree.DFloat(float64(numMinutes % 60)))
+	}
+	return nil
+}
+
+func extractStringFromTimeTZ(fromTime *tree.DTimeTZ, timeSpan string) (tree.Datum, error) {
+	if ret := extractTimezoneFromOffset(-fromTime.OffsetSecs, timeSpan); ret != nil {
+		return ret, nil
+	}
+	switch timeSpan {
 	case "epoch":
 		// Epoch should additionally add the zone offset.
 		seconds := float64(time.Duration(fromTime.TimeOfDay))*float64(time.Microsecond)/float64(time.Second) + float64(fromTime.OffsetSecs)
@@ -4557,6 +4567,22 @@ func dateToJulianDay(year int, month int, day int) int {
 	jd += 7834*month/256 + day
 
 	return jd
+}
+
+func extractStringFromTimestampTZ(
+	ctx *tree.EvalContext, fromTime time.Time, timeSpan string,
+) (tree.Datum, error) {
+	_, offsetSecs := fromTime.Zone()
+	if ret := extractTimezoneFromOffset(int32(offsetSecs), timeSpan); ret != nil {
+		return ret, nil
+	}
+
+	// time.Time's Year(), Month(), Day(), ISOWeek(), etc. all deal in terms
+	// of UTC, rather than as the timezone.
+	// Remedy this by assuming that the timezone is UTC (to prevent confusion)
+	// and offsetting time when using extractStringFromTimestamp.
+	pretendTime := fromTime.In(time.UTC).Add(time.Duration(offsetSecs) * time.Second)
+	return extractStringFromTimestamp(ctx, pretendTime, timeSpan)
 }
 
 func extractStringFromTimestamp(

--- a/pkg/sql/sem/builtins/builtins_test.go
+++ b/pkg/sql/sem/builtins/builtins_test.go
@@ -246,6 +246,80 @@ func TestLPadRPad(t *testing.T) {
 	}
 }
 
+func TestExtractStringFromTimestamp(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	utcPositiveOffset := time.FixedZone("otan happy time", 60*60*4+30*60)
+	utcNegativeOffset := time.FixedZone("otan sad time", -60*60*4-30*60)
+
+	testCases := []struct {
+		input    time.Time
+		timeSpan string
+
+		expected      tree.DFloat
+		expectedError string
+	}{
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "timezone", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "timezone_hour", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "timezone_minute", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "millennia", expected: 3},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "century", expected: 21},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "decade", expected: 201},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "year", expected: 2019},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "month", expected: 12},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "day", expected: 11},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "hour", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "minute", expected: 14},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "second", expected: 15.123456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "millisecond", expected: 15123.456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, time.UTC), timeSpan: "microsecond", expected: 15123456},
+
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "timezone", expected: 4*60*60 + 30*60},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "timezone_hour", expected: 4},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "timezone_minute", expected: 30},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "millennia", expected: 3},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "century", expected: 21},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "decade", expected: 201},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "year", expected: 2019},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "month", expected: 12},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "day", expected: 11},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "hour", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "minute", expected: 14},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "second", expected: 15.123456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "millisecond", expected: 15123.456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcPositiveOffset), timeSpan: "microsecond", expected: 15123456},
+
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "timezone", expected: -4*60*60 - 30*60},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "timezone_hour", expected: -4},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "timezone_minute", expected: -30},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "millennia", expected: 3},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "century", expected: 21},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "decade", expected: 201},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "year", expected: 2019},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "month", expected: 12},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "day", expected: 11},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "hour", expected: 0},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "minute", expected: 14},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "second", expected: 15.123456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "millisecond", expected: 15123.456},
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "microsecond", expected: 15123456},
+
+		{input: time.Date(2019, time.December, 11, 0, 14, 15, 123456000, utcNegativeOffset), timeSpan: "it's numberwang!", expectedError: "unsupported timespan: it's numberwang!"},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%s_%s", tc.timeSpan, tc.input.Format(time.RFC3339)), func(t *testing.T) {
+			datum, err := extractStringFromTimestampTZ(nil, tc.input, tc.timeSpan)
+			if tc.expectedError != "" {
+				assert.EqualError(t, err, tc.expectedError)
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, tc.expected, *(datum.(*tree.DFloat)))
+			}
+		})
+	}
+}
+
 func TestExtractStringFromTimeTZ(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

